config/compat.go

```diff
@@ -0,0 +1,172 @@
+package config
+
+import (
+	"fmt"
+	"io"
+	"os"
+
+	"gopkg.in/yaml.v3"
+)
+
+type configV112 struct {
+	Name           string `yaml:"name,omitempty"`
+	Directory      string `yaml:"directory,omitempty"`
+	RecoveryPhrase string `yaml:"recoveryPhrase,omitempty"`
+	AutoOpenWebUI  bool   `yaml:"autoOpenWebUI,omitempty"`
+
+	HTTP struct {
+		Address  string `yaml:"address,omitempty"`
+		Password string `yaml:"password,omitempty"`
+	} `yaml:"http,omitempty"`
+	Consensus struct {
+		GatewayAddress string   `yaml:"gatewayAddress,omitempty"`
+		Bootstrap      bool     `yaml:"bootstrap,omitempty"`
+		Peers          []string `yaml:"peers,omitempty"`
+	} `yaml:"consensus,omitempty"`
+	Explorer struct {
+		Disable bool   `yaml:"disable,omitempty"`
+		URL     string `yaml:"url,omitempty"`
+	} `yaml:"explorer,omitempty"`
+	RHP2 struct {
+		Address string `yaml:"address,omitempty"`
+	} `yaml:"rhp2,omitempty"`
+	RHP3 struct {
+		TCPAddress       string `yaml:"tcp,omitempty"`
+		WebSocketAddress string `yaml:"websocket,omitempty"`
+		CertPath         string `yaml:"certPath,omitempty"`
+		KeyPath          string `yaml:"keyPath,omitempty"`
+	} `yaml:"rhp3,omitempty"`
+	Log struct {
+		// Path is the directory to store the hostd.log file.
+		// Deprecated: use File.Path instead.
+		Path   string `yaml:"path,omitempty"`
+		Level  string `yaml:"level,omitempty"` // global log level
+		StdOut struct {
+			Level      string `yaml:"level,omitempty"` // override the stdout log level
+			Enabled    bool   `yaml:"enabled,omitempty"`
+			Format     string `yaml:"format,omitempty"`
+			EnableANSI bool   `yaml:"enableANSI,omitempty"` //nolint:tagliatelle
+		} `yaml:"stdout,omitempty"`
+		File struct {
+			Enabled bool   `yaml:"enabled,omitempty"`
+			Level   string `yaml:"level,omitempty"` // override the file log level
+			Format  string `yaml:"format,omitempty"`
+			// Path is the path of the log file.
+			Path string `yaml:"path,omitempty"`
+		} `yaml:"file,omitempty"`
+	} `yaml:"log,omitempty"`
+}
+
+// updateConfigV112 updates the config file from v1.1.2 to the latest version.
+// It returns an error if the config file cannot be updated.
+func updateConfigV112(fp string, r io.Reader, cfg *Config) error {
+	dec := yaml.NewDecoder(r)
```

Comment:
are you purposefully not setting `KnownFields` here?

Reply:
Changed my mind. I think you're right, we should fail if the config is invalid.